File: cpp/045.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int start = 0, end = 0, res = 0, size = nums.size();
        for (; start < size - 1;) {
            ++res;
            int next = 0;
            for (int i = start; i <= end; ++i) {
                next = max(next, i + nums[i]);
                if (next >= size - 1)
                    return res;
            }
            start = end + 1;
            end = next;
        }
        return res;
    }
};

File: go/045.go
func jump(nums []int) int {
    start, end, res, size := 0, 0, 0, len(nums)
    for start < size-1 {
        res++
        next := 0
        for i := start; i <= end; i++ {
            next = max(next, i+nums[i])
            if next >= size-1 {
                return res
            }
        }
        start, end = end+1, next
    }
    return res
}

func max(a, b int) int {
    if a < b {
        return b
    }
    return a
}
